src/System.Text.Json/src/System/Text/Json/Reader/JsonReaderHelper.Date.cs

```diff
@@ -0,0 +1,503 @@
+            {
+                sourceIndex += 1;
+            }
+
+            if (!TryGetNextTwoDigits(source.Slice(start: sourceIndex, length: 2), out offsetMinutes))
+            {
+                goto ReturnFalse;
+            }
+            sourceIndex += 2;
+
+            // We now have YYYY-MM-DDThh:mm:ss.s+|-hh[:]mm
+            bytesConsumed = sourceIndex;
+
+        FinishedParsing:
+            if ((offsetToken != JsonConstants.UtcOffsetToken) && (offsetToken != JsonConstants.Plus) && (offsetToken != JsonConstants.Minus))
+            {
+                if (!TryCreateDateTimeOffsetInterpretingDataAsLocalTime(year: year, month: month, day: day, hour: hour, minute: minute, second: second, fraction: fraction, out value))
+                {
+                    goto ReturnFalse;
+                }
+
+                kind = DateTimeKind.Unspecified;
+                return true;
+            }
+
+            if (offsetToken == JsonConstants.UtcOffsetToken)
+            {
+                // Same as specifying an offset of "+00:00", except that DateTime's Kind gets set to UTC rather than Local
+                if (!TryCreateDateTimeOffset(year: year, month: month, day: day, hour: hour, minute: minute, second: second, fraction: fraction, offsetNegative: false, offsetHours: 0, offsetMinutes: 0, out value))
+                {
+                    goto ReturnFalse;
+                }
+
+                kind = DateTimeKind.Utc;
+                return true;
+            }
+
+            Debug.Assert(offsetToken == JsonConstants.Plus || offsetToken == JsonConstants.Minus);
+
+            if (!TryCreateDateTimeOffset(year: year, month: month, day: day, hour: hour, minute: minute, second: second, fraction: fraction, offsetNegative: offsetToken == JsonConstants.Minus, offsetHours: offsetHours, offsetMinutes: offsetMinutes, out value))
+            {
+                goto ReturnFalse;
+            }
+
+            kind = DateTimeKind.Local;
+            return true;
+
+        ReturnFalse:
+            value = default;
+            bytesConsumed = 0;
+            kind = default;
+            return false;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool TryGetNextTwoDigits(ReadOnlySpan<byte> source, out int value)
+        {
+            Debug.Assert(source.Length == 2);
+
+            uint digit1 = source[0] - (uint)'0';
+            uint digit2 = source[1] - (uint)'0';
+
+            if (digit1 > 9 || digit2 > 9)
+            {
+                value = default;
+                return false;
+            }
+
+            value = (int)(digit1 * 10 + digit2);
+            return true;
+        }
+
+        /// <summary>
+        /// Overflow-safe DateTimeOffset factory.
+        /// </summary>
+        private static bool TryCreateDateTimeOffset(DateTime dateTime, bool offsetNegative, int offsetHours, int offsetMinutes, out DateTimeOffset value)
```

Comment:
If these methods are borrowed verbatim, please add a comment stating from where.

Alternatively, we could consider including [Utf8Parser.Date.Helpers.cs](https://github.com/dotnet/corefx/blob/9c68db7fb016c6c9ae4d0f6152798d7ab1e38a37/src/Common/src/CoreLib/System/Buffers/Text/Utf8Parser/Utf8Parser.Date.Helpers.cs to the project rather than copying the source.

Reply:
Added a comment stating from where.